ulib/FStar.Seq.Permutation.fst

```diff
@@ -491,11 +491,25 @@ let rec foldm_snoc_split' #c #eq (cm: CE.cm c eq)
       let nk' = nk - 1 in
       (* here's the nasty bit with where we have to massage the proof from the induction hypothesis *)
       let ih
-        : squash ((foldm_snoc cm (init (range_count n0 nk') (init_func_from_expr (func_sum cm expr1 expr2) n0 nk')) `eq.eq`
-              cm.mult (foldm_snoc cm (init (range_count n0 nk') (init_func_from_expr expr1 n0 nk')))
-                      (foldm_snoc cm (init (range_count n0 nk') (init_func_from_expr expr2 n0 nk')))))
+        : squash ((foldm_snoc cm (init (range_count n0 nk') (init_func_from_expr #_ #n0 #nk' (func_sum #(in_between n0 nk') cm expr1 expr2) n0 nk')) `eq.eq`
+              cm.mult (foldm_snoc cm (init (range_count n0 nk') (init_func_from_expr #_ #n0 #nk' expr1 n0 nk')))
+                      (foldm_snoc cm (init (range_count n0 nk') (init_func_from_expr #_ #n0 #nk' expr2 n0 nk')))))
         = foldm_snoc_split' cm n0 nk' expr1 expr2
       in
+      calc (==) {
+        subfold;
+      == { }
+        foldm_snoc cm subseq;
+      == { assert (Seq.equal subseq
+                             (init (range_count n0 nk')
+                                   (init_func_from_expr #_ #n0 #nk'
+                                      (func_sum #(in_between n0 nk') cm expr1 expr2) n0 nk'))) }
+        foldm_snoc cm (init (range_count n0 nk')
+                            (init_func_from_expr #_ #n0 #nk'
+                               (func_sum #(in_between n0 nk') cm expr1 expr2) n0 nk'));
+      };
+      assert (Seq.equal subseq_r1 (init (range_count n0 nk') (init_func_from_expr #_ #n0 #nk' expr1 n0 nk')));
+      assert (Seq.equal subseq_r2 (init (range_count n0 nk') (init_func_from_expr #_ #n0 #nk' expr2 n0 nk')));
```

Comment:
I guess this is a fresh proof, so it's still fragile? Any other explanation for regression?

Reply:
One thing that I found tricky in this code is the implicit instantiation. The changes that I made are mostly related to explicitly instantiating the implicits, specifically the `nk'` ones. Consider for example:

```
let func_sum #a #c #eq (cm: CE.cm c eq) (f g: a → c) ...
```

And we have `expr1 expr2:in_between n0 nk -> c`.

Now, within the recursive call `func_sum cm expr1 expr2`, we are expecting the `#a` implicit to be instantiated as `in_between n0 nk'`, where `nk' = nk - 1`, which seems a bit tricky. I think `expr1` and `expr2` can be typed `in_between n0 nk' -> c`, but I thought a different order of implicit resolution could lead to problems.

Similar argument for `init_func_expr` etc.

That was my analysis from the incorrect implicit instantiations that I saw. I guessed that may be code changes in the PR somehow resulted in slightly different order of implicit resolution leading to it. But I did not go into more details.